File: app/src/main/java/com/doapps/habits/fragments/ProfileEditFragment.java
package com.doapps.habits.fragments;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.text.InputType;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.doapps.habits.BuildConfig;
import com.doapps.habits.R;
import com.doapps.habits.activity.EditPhotoActivity;
import com.doapps.habits.activity.MainActivity;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.FacebookSdk;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.EmailAuthProvider;
import com.google.firebase.auth.FacebookAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserInfo;
import com.google.firebase.auth.UserProfileChangeRequest;

@SuppressWarnings("FeatureEnvy")
public class ProfileEditFragment extends Fragment {

    private String inputPassword = "";
    private static final String TAG = ProfileEditFragment.class.getSimpleName();
    private FloatingActionButton fab;
    private ImageView avatar;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View result = inflater.inflate(R.layout.fragment_edit_profile, container, false);
        fab = (FloatingActionButton) getActivity().findViewById(R.id.fab);
        avatar = (ImageView) getActivity().findViewById(R.id.avatarImage);
        setUpPhotoEdition(avatar, getActivity(), this);

        fab.setImageResource(R.drawable.ic_check_white_24dp);
        fab.setOnClickListener(view -> {
                    FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();

                    if (user != null && user.getEmail() != null) {
                        String email =
                                String.valueOf(((TextView) result.findViewById(R.id.edit_email)).getText()).toLowerCase();
                        String name =
                                String.valueOf(((TextView) result.findViewById(R.id.edit_name)).getText());

                        if (!isUpdated(user, email, name)) { // same info
                            getFragmentManager().beginTransaction().remove(this).commit();
                            Log.i("EditProfile", "Nothing to update");
                            return;
                        }
                        if (MainActivity.isFacebook(user)) {
                            // TODO: 14/07/2016 change to working solution
                            CallbackManager callbackManager =
                                    ((MainActivity) getActivity()).getCallbackManager();
                            FacebookSdk.sdkInitialize(getContext().getApplicationContext());
                            LoginManager.getInstance().registerCallback(callbackManager,
                                    new FacebookCallback<LoginResult>() {
                                        @Override
                                        public void onSuccess(LoginResult result) {
                                            if (BuildConfig.DEBUG) {
                                                Log.d(TAG, "facebook:onSuccess:" + result);
                                            }
                                            AuthCredential credential =
                                                    FacebookAuthProvider.getCredential(result.getAccessToken().getToken());
                                            user.reauthenticate(credential)
                                                    .addOnCompleteListener(task -> Log.d(TAG, "User re-authenticated."));
                                            updateUser(getActivity(), user, name, email);
                                        }

                                        @Override
                                        public void onCancel() {
                                            Log.d(TAG, "facebook:onCancel");
                                            // ...
                                        }

                                        @Override
                                        public void onError(FacebookException error) {
                                            Log.d(TAG, "facebook:onError", error);
                                        }
                                    });

                            getFragmentManager().beginTransaction().remove(this).commit();
                        } else {
                            getPasswordFromUser();
                            // Prompt the user to re-provide their sign-in credentials
                            AuthCredential credential = EmailAuthProvider
                                    .getCredential(user.getEmail(), inputPassword);
                            user.reauthenticate(credential)
                                    .addOnCompleteListener(task -> Log.d(TAG, "User re-authenticated."));
                            if (inputPassword.isEmpty()) { // entered password empty
                                Toast.makeText(getContext().getApplicationContext(),
                                        "Please enter correct password", Toast.LENGTH_SHORT)
                                        .show();
                                getPasswordFromUser();
                                return;
                            }
                        }

                        updateUser(getActivity(), user, name, email);
                        getFragmentManager().beginTransaction().remove(this).commit();
                    } else {
                        Log.e(TAG, "User is null");
                    }
                }

        );


        return result;
    }

    private static boolean isUpdated(UserInfo user, String email, String name) {
        boolean nameUpdated = !name.isEmpty() && !name.equals(user.getDisplayName());
        boolean emailUpdated = !email.isEmpty() && !email.equals(user.getEmail());
        return nameUpdated || emailUpdated;
    }

    private void getPasswordFromUser() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle("Please re-enter your password");

        // Set up the input
        EditText input = new EditText(getActivity());
        // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
        input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        builder.setView(input);

        // Set up the buttons
        builder.setPositiveButton("OK", (dialog, which) -> inputPassword = input.getText().toString());
        builder.setNegativeButton("Cancel", (dialog, which) -> {
            getFragmentManager().beginTransaction().remove(this).commit();
        });

        builder.show();
    }

    private static void updateUser(Activity activity, FirebaseUser user,
                                   String name, String email) {
        TextView nameView =
                (TextView) activity.findViewById(R.id.name);
        TextView emailView =
                (TextView) activity.findViewById(R.id.email);

        if (!name.isEmpty() && !name.equals(user.getDisplayName())) {
            UserProfileChangeRequest profileUpdates =
                    new UserProfileChangeRequest.Builder()
                            .setDisplayName(name)
                            .build();

            user.updateProfile(profileUpdates)
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            NavigationView navigationView =
                                    (NavigationView) activity.findViewById(R.id.navigationView);

                            ((TextView) navigationView
                                    .getHeaderView(0).findViewById(R.id.name_info))
                                    .setText(user.getDisplayName());

                            nameView.setText(user.getDisplayName());

                            Log.d(TAG, "User profile updated.");
                        }
                    });
        }
        if (!email.isEmpty() && !email.equals(user.getEmail())) {
            user.updateEmail(email)
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            NavigationView navigationView =
                                    (NavigationView) activity.findViewById(R.id.navigationView);

                            ((TextView) navigationView
                                    .getHeaderView(0).findViewById(R.id.email_info))
                                    .setText(user.getEmail());

                            emailView.setText(user.getEmail());

                            Log.d(TAG, "User email address updated.");
                        }
                    });
        }

    }

    private static void setUpPhotoEdition(ImageView avatar, Activity activity, Fragment fragment) {
        avatar.setImageResource(R.drawable.ic_photo_plus);
        if (avatar.getBackground() != null) {
            avatar.getBackground().setColorFilter(Color.GRAY, PorterDuff.Mode.MULTIPLY);
        }
        avatar.setOnClickListener(view -> {
            fragment.getParentFragment().getChildFragmentManager().beginTransaction()
                    .remove(fragment).commit();
            Intent intent = new Intent(activity, EditPhotoActivity.class);
            activity.startActivity(intent);
        });
    }


    private static void removePhotoEdition(ImageView avatar) {
        avatar.setImageResource(R.drawable.fix);
        if (avatar.getBackground() != null) {
            avatar.getBackground().clearColorFilter();
        }
        avatar.setOnClickListener(null);
    }

    @Override
    public void onPause() {
        removePhotoEdition(avatar);
        // Set up fab back
        fab.setImageResource(R.drawable.ic_edit_white_24dp);
        fab.setOnClickListener(v -> {
            if (!ProfileFragment.getEditorOpened()[0]) {
                getParentFragment().getChildFragmentManager()
                        .beginTransaction()
                        .replace(R.id.userInfo, new ProfileEditFragment())
                        .commit();
                ProfileFragment.getEditorOpened()[0] = true;
            }
        });

        // Sets editors state to close
        ProfileFragment.getEditorOpened()[0] = false;
        // Close keyboard
        ((MainActivity) getActivity()).closeImm();
        super.onPause();
    }
}

File: app/src/main/java/com/doapps/habits/fragments/ProfileEditFragment.kt
package com.doapps.habits.fragments

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.support.v7.app.AlertDialog
import android.text.InputType
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.doapps.habits.BuildConfig
import com.doapps.habits.R
import com.doapps.habits.activity.EditPhotoActivity
import com.doapps.habits.activity.MainActivity
import com.doapps.habits.data.AvatarData
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.google.firebase.auth.*

class ProfileEditFragment : Fragment() {
  private var inputPassword = ""
  private var fab: FloatingActionButton? = null
  private lateinit var plus: ImageView
  private val user: FirebaseUser? = FirebaseAuth.getInstance().currentUser

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    // Inflate the layout for this fragment
    val result = inflater.inflate(R.layout.fragment_edit_profile, container, false)
    fab = activity!!.findViewById(R.id.fab)
    plus = activity!!.findViewById(R.id.plus_overlay)
    if (plus.imageAlpha == 255) setUpPhotoEdition(plus, activity, this)

    fab!!.setImageResource(R.drawable.ic_check_white_24dp)
    fab!!.setOnClickListener {
      if (user != null && user.email != null) {
        val email = (result.findViewById<View>(R.id.edit_email) as TextView).text.toString().toLowerCase()
        val name = (result.findViewById<View>(R.id.edit_name) as TextView).text.toString()

        if (!isUpdated(user, email, name)) { // same info
          fragmentManager!!.beginTransaction().remove(this).commit()
          Log.i(TAG, "Nothing to update")
          return@setOnClickListener
        }
        if (MainActivity.isFacebook(user)) {
          val callbackManager = (activity as MainActivity).callbackManager
          LoginManager.getInstance().registerCallback(callbackManager,
              object : FacebookCallback<LoginResult> {
                override fun onSuccess(result: LoginResult) {
                  if (BuildConfig.DEBUG) {
                    Log.d(TAG, "facebook:onSuccess:" + result)
                  }
                  val credential = FacebookAuthProvider.getCredential(result.accessToken.token)
                  user.reauthenticate(credential).addOnCompleteListener { Log.d(TAG, "User re-authenticated.") }
                  updateUser(activity, user, name, email)
                }

                override fun onCancel() {
                  Log.d(TAG, "facebook:onCancel")
                }

                override fun onError(error: FacebookException) {
                  Log.d(TAG, "facebook:onError", error)
                }
              })

          fragmentManager!!.beginTransaction().remove(this).commit()
        } else {
          getPasswordFromUser()
          // Prompt the user to re-provide their sign-in credentials
          val credential = EmailAuthProvider
              .getCredential(user.email!!, inputPassword)
          user.reauthenticate(credential).addOnCompleteListener { Log.d(TAG, "User re-authenticated.") }
          if (inputPassword.isEmpty()) { // entered password empty
            Toast.makeText(context!!.applicationContext,
                "Please enter correct password", Toast.LENGTH_SHORT)
                .show()
            getPasswordFromUser()
            return@setOnClickListener
          }
        }

        updateUser(activity, user, name, email)
        fragmentManager!!.beginTransaction().remove(this).commit()
      } else {
        Log.e(TAG, "User is null")
      }
    }

    return result
  }

  private fun getPasswordFromUser() {
    val builder = AlertDialog.Builder(activity!!)
    builder.setTitle("Please re-enter your password")

    // Set up the input
    val input = EditText(activity)
    // Specify the type of input expected; this, for example, sets the input as a password, and will mask the text
    input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
    builder.setView(input)

    // Set up the buttons
    builder.setPositiveButton("OK") { _, _ -> inputPassword = input.text.toString() }
    builder.setNegativeButton("Cancel") { _, _ -> fragmentManager!!.beginTransaction().remove(this).commit() }

    builder.show()
  }

  override fun onPause() {
    if (AvatarData.hasAvatar(context!!)) {
      plus.setOnClickListener(null)
      plus.visibility = View.GONE
    }
    // Set up fab back
    fab!!.setImageResource(R.drawable.ic_edit_white_24dp)
    fab!!.setOnClickListener {
      if (!ProfileFragment.editorOpened[0]) {
        parentFragment!!.childFragmentManager
            .beginTransaction()
            .replace(R.id.user_info, ProfileEditFragment())
            .commit()
        ProfileFragment.editorOpened[0] = true
      }
    }

    // Sets editors state to close
    ProfileFragment.editorOpened[0] = false
    // Close keyboard
    (activity as MainActivity).closeImm()
    super.onPause()
  }

  companion object {

    /**
     * TAG is defined for logging errors and debugging information
     */
    private val TAG = ProfileEditFragment::class.java.simpleName

    private fun isUpdated(user: UserInfo, email: String, name: String): Boolean {
      val nameUpdated = !name.isEmpty() && name != user.displayName
      val emailUpdated = !email.isEmpty() && email != user.email
      return nameUpdated || emailUpdated
    }

    private fun updateUser(activity: Activity?, user: FirebaseUser?,
                           name: String, email: String) {
      val nameView = activity!!.findViewById<TextView>(R.id.name)
      val emailView = activity.findViewById<TextView>(R.id.email)

      if (!name.isEmpty() && name != user!!.displayName) {
        val profileUpdates = UserProfileChangeRequest.Builder()
            .setDisplayName(name)
            .build()

        user.updateProfile(profileUpdates)
            .addOnCompleteListener { task ->
              if (task.isSuccessful) {
                val navigationView = activity.findViewById<NavigationView>(R.id.navigationView)

                (navigationView
                    .getHeaderView(0).findViewById<View>(R.id.name_info) as TextView).text = user.displayName

                nameView.text = user.displayName

                Log.d(TAG, "User profile updated.")
              }
            }
      }
      if (!email.isEmpty() && email != user!!.email) {
        user.updateEmail(email)
            .addOnCompleteListener { task ->
              if (task.isSuccessful) {
                val navigationView = activity.findViewById<NavigationView>(R.id.navigationView)

                (navigationView
                    .getHeaderView(0).findViewById<View>(R.id.email_info) as TextView).text = user.email

                emailView.text = user.email

                Log.d(TAG, "User email address updated.")
              }
            }
      }

    }

    private fun setUpPhotoEdition(plus: ImageView, activity: Activity?, fragment: Fragment) {
      plus.visibility = View.VISIBLE
      plus.setOnClickListener {
        fragment.parentFragment!!.childFragmentManager.beginTransaction()
            .remove(fragment).commit()
        val intent = Intent(activity, EditPhotoActivity::class.java)
        activity!!.startActivity(intent)
      }
    }
  }
}
